src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ConsiderPassingBaseTypesAsParameters.cs

```diff
@@ -0,0 +1,343 @@
+                            paramType,
+                            mostDerived));
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Record the most derived usage type (and all usage types) to use in 
+        /// our evaluation at the end of the method. Note that setting usage type
+        /// to the parameter's type effectively shuts down the analysis, because
+        /// it will always be the most derived type.
+        /// </summary>
+        private void RecordParameterUse(
+            IParameterSymbol parameter,
+            ITypeSymbol usageType,
+            TMapParameterToType mostDerivedTypeFor,
+            TMapParameterToTypeSet allDerivedTypesFor)
+        {
+            var paramType = parameter.Type;
+
+            if (parameter.IsThis
+                || parameter.IsImplicitlyDeclared 
+                || paramType.IsValueType
+                || paramType.SpecialType != SpecialType.None // ignore all the special types
+                || paramType.IsValueType
+                )
+            {
+                return;
+            }
+
+            // if the usage type is not compatible with the parameter, shutdown
+            // the analysis by making usage type the parameter type
+            if (!IsAssignableTo(paramType, usageType))
+            {
+                usageType = paramType;
+            }
+
+            // Record the most derived usage type
+            ITypeSymbol mostDerivedTypeSoFar;
+            if (!mostDerivedTypeFor.TryGetValue(parameter, out mostDerivedTypeSoFar)
+                || IsAssignableTo(usageType, mostDerivedTypeSoFar))
+            {
+                mostDerivedTypeFor[parameter] = usageType;
+            }
+
+            // Save all the usage types, so we can later verify that the most derived type
+            // is assignable to them all (i.e. in case we get a class that implements two 
+            // unrelated interfaces)
+            HashSet<ITypeSymbol> derivedTypes = null;
+            if (!allDerivedTypesFor.TryGetValue(parameter, out derivedTypes))
+            {
+                derivedTypes = new HashSet<ITypeSymbol>();
+                allDerivedTypesFor[parameter] = derivedTypes;
+            }
+            if (!derivedTypes.Contains(usageType))
+            {
+                derivedTypes.Add(usageType);
+            }
+        }
+
+        /// <summary>
+        /// Utility to determine if the IOperation is a parameter reference. 
+        /// Returns null if it is not.
+        /// </summary>
+        static IParameterSymbol GetParameterSymbol(IOperation value)
+        {
+            if (value?.Kind == OperationKind.ConversionExpression)
+            {
+                value = ((IConversionExpression)value).Operand;
+            }
+
+            return (value as IParameterReferenceExpression)?.Parameter;
+        }
+
+        static bool IsAssignableTo(ITypeSymbol derivedType, ITypeSymbol baseType)
```

Comment:
I got this function from one of John Hamby's examples. I can move it to a helper, or use Larry's if it is equivalent. I'll investigate